# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv("../data/1_2_data_prep_data.csv")
except:
    data = pd.read_csv("https://code.s3.yandex.net/datasets/data.csv")

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data["income_type"].unique():
    data.loc[
        (data["income_type"] == t) & (data["total_income"].isna()), "total_income"
    ] = data.loc[(data["income_type"] == t), "total_income"].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data["days_employed"] = data["days_employed"].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby("income_type")["days_employed"].agg("median")

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data["children"].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data["children"] != -1) & (data["children"] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data["children"].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data["income_type"].unique():
    data.loc[
        (data["income_type"] == t) & (data["days_employed"].isna()), "days_employed"
    ] = data.loc[(data["income_type"] == t), "days_employed"].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data["total_income"] = data["total_income"].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data["education"] = data["education"].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

np.int64(71)

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return "E"
        elif 30001 <= income <= 50000:
            return "D"
        elif 50001 <= income <= 200000:
            return "C"
        elif 200001 <= income <= 1000000:
            return "B"
        elif income >= 1000001:
            return "A"
    except:
        pass

In [18]:
data["total_income_category"] = data["total_income"].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data["purpose"].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if "автом" in row:
            return "операции с автомобилем"
        elif "жил" in row or "недвиж" in row:
            return "операции с недвижимостью"
        elif "свад" in row:
            return "проведение свадьбы"
        elif "образов" in row:
            return "получение образования"
    except:
        return "нет категории"

In [21]:
data["purpose_category"] = data["purpose"].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

<a class="ancor" id="0-bullet"></a>
### Оглавление

* [Влияние количества детей](#1-bullet)
* [Влияние семейного положения](#2-bullet)
* [Влияние уровня дохода](#3-bullet)
* [Влияние цели кредита](#4-bullet)
* [Анализ причин возникновения пропусков данных](#5-bullet)
* [Обсуждение использования медианы для заполнения пропусков данных](#6-bullet)
* [Выводы](#7-bullet)

<a class="ancor" id="1-bullet"></a>
#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

Для начала посмотрим размеры групп клиентов с разным количеством детей

In [22]:
# В дальнейшем будем использовать named aggregation, чтобы получить сразу размер групп и их процент от общего количества клиентов
data.groupby("children").agg(
    value=("children", "count"),
    percent=("children", lambda x: f"{x.count()/len(data)*100:.2f}%"),
)

value percent
children               
0         14091  66.06%
1          4808  22.54%
2          2052   9.62%
3           330   1.55%
4            41   0.19%
5             9   0.04%

Суммарное количество клиентов с 4 и более детьми составляет менее 0.25% от общего количества клиентов, поэтому данные для таких клиентов могут быть статистически недостоверными и их учитывать не будем.

Количество клиентов с 3 детьми также довольно низкое, оставим их для анализа, но будем иметь ввиду, что данные для этой категории не особо надёжны.

In [23]:
data_childs = data[data["children"] < 4]
# Проверим, что данные действительно содержат нужные категории
data_childs["children"].unique()

array([1, 0, 3, 2])

Посмотрим распределение доли просроченных кредитов от количества детей

In [24]:
def debt_data_by_group(group_col: str, size_thr: float = 0.005) -> None:
    """
    Функция для построения сводной таблицы просрочки для в зависимости от группы разбиения

    params
    ----------
    group_col: str
        Название колонки с группами
    size_thr: float
        Пороговое значение размера групп. Группы с размером ниже порога не будут учитываться.
        По умолчанию пороговое значение равно 0.5 %.
    """

    # Собираем данные по размерам групп
    grouped_data = data.groupby(group_col).agg(
        group_size=(group_col, "count"),
        group_size_percent=(group_col, lambda x: f"{x.count()/len(data)*100:.2f}"),
    )

    # Удалим из анализа группы с размером ниже порога, предварительно сообщив об этом
    if len(
        small_groups := grouped_data[
            grouped_data["group_size_percent"].astype(float) <= size_thr * 100
        ]
    ):
        print(
            f"Размеры групп клиентов с {group_col} = {small_groups.index.values} "
            + f"слишком малы для анализа и не будут использоваться."
        )
    grouped_data = grouped_data[
        grouped_data["group_size_percent"].astype(float) > size_thr * 100
    ]

    # Не забываем отфильтровать и исходные данные
    data_filtered = data[data[group_col].isin(grouped_data.index)]

    # Собираем данные по распределению просрочки кредитов в группах
    max_debt_fraction = data_filtered.groupby(group_col)["debt"].mean().max()
    grouped_data = grouped_data.join(
        data_filtered.groupby(group_col).agg(
            debt_fraction=("debt", "mean"),
            relative_debt_percent=(
                "debt",
                lambda x: f"{(x.mean()/max_debt_fraction - 1)*100:.2f}",
            ),
        )
    )
    display(grouped_data)

In [25]:
debt_data_by_group("children")

Размеры групп клиентов с children = [4 5] слишком малы для анализа и не будут использоваться.


group_size group_size_percent  debt_fraction relative_debt_percent
children                                                                    
0              14091              66.06       0.075438                -20.21
1               4808              22.54       0.092346                 -2.32
2               2052               9.62       0.094542                  0.00
3                330               1.55       0.081818                -13.46

Максимальная доля просроченных кредитов наблюдается у клиентов с 1 и 2 детьми и составляет чуть более 9 %. Для клиентов с 3 детьми вероятность просрочки составляет 8.2 %, а для клиентов без детей 7.5 %, что на 13 % и 20 % меньше, чем для самых проблемных клиентов.

Выводы выше основаны на усреднённом "здравом смысле", но можно получить и более строгие выводы.

К счастью, в нашем случае групп мало и можно легко сделать тест статистической значимости для всех попарных комбинаций групп клиентов с разными количествами детей.

Напишем вспомогательную функцию.

In [26]:
from typing import Literal, Optional, cast
from itertools import combinations

# Нужно установить библиотеку scipy
from scipy.stats import ttest_ind, chi2_contingency


def statistical_significance(
    df: pd.DataFrame,
    group_col: str,
    value_col: str,
    p_thr: float,
    method: Literal["welch", "chi_squared"],
    groups: Optional[list[str]] = None,
) -> None:
    """
    Функция для попарного сравнения статистической значимости значений внутри групп.

    params
    ----------
    df: pd.DataFrame
        Датафрейм с данными
    group_col: str
        Название колонки с группами
    value_col: str
        Название колонки со значениями
    p_thr: float
        Пороговое p-значение для определения статистической значимости
    method: Literal['welch','chi_squared']
        Методика подсчёта статистической значимости
    groups: list[str]|None
        Список групп для сравнения. Если None, то сравниваются все группы
    """
    # Группы сравнения
    test_groups = groups or df[group_col].unique()

    pairwise_results = []
    # Перебираем все попарные комбинации
    for group1, group2 in combinations(test_groups, 2):
        group1_data = cast(pd.Series, df.loc[df[group_col] == group1, value_col])
        group2_data = cast(pd.Series, df.loc[df[group_col] == group2, value_col])
        if method == "welch":
            # Считаем t-статистику и p-value. Популяционные дисперсии нам неизвестны, поэтому предполагаем, что они разные (equal_var=False)
            stat, pvalue = ttest_ind(group1_data, group2_data, equal_var=False)
            pairwise_results.append((group1, group2, stat, pvalue))
        elif method == "chi_squared":
            # Создаём таблицу сопряженности
            contingency_table = pd.crosstab(
                df.loc[df[group_col].isin([group1, group2]), group_col],
                df.loc[df[group_col].isin([group1, group2]), value_col],
            )
            # Выполняем хи-квадрат тест
            chi2, pvalue, _, _ = chi2_contingency(contingency_table)
            pairwise_results.append((group1, group2, chi2, pvalue))
        else:
            raise ValueError("Метод должен быть 'welch' или 'chi_squared'")

    # Печатаем результаты
    for result in pairwise_results:
        print(
            f"Для клиентов с {group_col} = {result[0]} и {result[1]}  p-значение = {result[3]:.4f}.",
            end="",
        )
        if result[3] < p_thr:
            print("  Различия статистически значимы.")
        else:
            print("  Различия статистически НЕ значимы.")

Проверим статистическую значимость полученных результатов.

Порогом статистической значимости сейчас и далее будем считать p-value=0.05 (типичное значение).

In [27]:
statistical_significance(data_childs, "children", "debt", 0.05, "chi_squared")

Для клиентов с children = 1 и 0  p-значение = 0.0002.  Различия статистически значимы.
Для клиентов с children = 1 и 3  p-значение = 0.5875.  Различия статистически НЕ значимы.
Для клиентов с children = 1 и 2  p-значение = 0.8093.  Различия статистически НЕ значимы.
Для клиентов с children = 0 и 3  p-значение = 0.7429.  Различия статистически НЕ значимы.
Для клиентов с children = 0 и 2  p-значение = 0.0029.  Различия статистически значимы.
Для клиентов с children = 3 и 2  p-значение = 0.5240.  Различия статистически НЕ значимы.


Таким образом, статистически значимо различаются вероятности просрочки между клиентами без детей и клиентами с 1 или 2 детьми.

Таким образом, результаты более строгого аналза вполне хорошо согласуются с результами, полученными с использованием "здравого смысла". В дальнейшем будем использовать этот более строгий подход только в случаях, когда статистическая значимость различий неочевидна на первый взгляд.

**Вывод:**

Гипотеза о том, что вероятность возврата кредита в срок зависит от количества детей обоснована для клиентов с количеством детей менее 4.

Среди таких клиентов самыми надежными являются клиенты без детей и с 3 детьми (вероятность просрочки около 8 %), а ненадежными являются клиенты с 1 и 2 детьми (вероятность просрочки более 9 %). Остальные различия в результатах статистически НЕ значимы.

Однако, стоит отметить, что для клиентов с 3 детьми количество данных относительно невелико (330 записей), поэтому лучше проверить достоверность результатов для этой группе на большем датасете.

Для клиентов с 4 и более детьми достоверных выводов сделать нельзя ввиду малочисленности таких клиентов (суммарно менее 0.25 %).

[Оглавление](#0-bullet)

<a class="ancor" id="2-bullet"></a>
#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

Посмотрим какие варианты семейного положения представлены в анализируемых данных.

In [28]:
data["family_status"].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

Посмотрим распределение средней просрочки от семейного статуса

In [29]:
debt_data_by_group("family_status")

group_size group_size_percent  debt_fraction  \
family_status                                                         
Не женат / не замужем        2796              13.11       0.097639   
в разводе                    1189               5.57       0.070648   
вдовец / вдова                951               4.46       0.066246   
гражданский брак             4134              19.38       0.093130   
женат / замужем             12261              57.48       0.075606   

                      relative_debt_percent  
family_status                                
Не женат / не замужем                  0.00  
в разводе                            -27.64  
вдовец / вдова                       -32.15  
гражданский брак                      -4.62  
женат / замужем                      -22.57

Рзамеры всех групп достаточны для анализа.

Максимальная доля просроченных кредитов наблюдается у не женатых/не замужних клиентов, а также людей, состоящих в гражданском браке (более 9 %), остальные категории клиентов на 20-30 % более надежны.

Различия в статистической значимости между группами "гражданский брак" и "Не женат / не замужем" не очевидны, воспользуемся тестом Уэлча.

In [30]:
statistical_significance(
    data,
    "family_status",
    "debt",
    0.05,
    "chi_squared",
    ["Не женат / не замужем", "гражданский брак"],
)

Для клиентов с family_status = Не женат / не замужем и гражданский брак  p-значение = 0.5576.  Различия статистически НЕ значимы.


Различия в статистической значимости между группами "в разводе", "вдовец / вдова" и "женат / замужем" также не очевидны, воспользуемся тестом Уэлча.

In [31]:
statistical_significance(
    data,
    "family_status",
    "debt",
    0.05,
    "chi_squared",
    ["вдовец / вдова", "женат / замужем", "в разводе"],
)

Для клиентов с family_status = вдовец / вдова и женат / замужем  p-значение = 0.3211.  Различия статистически НЕ значимы.
Для клиентов с family_status = вдовец / вдова и в разводе  p-значение = 0.7535.  Различия статистически НЕ значимы.
Для клиентов с family_status = женат / замужем и в разводе  p-значение = 0.5745.  Различия статистически НЕ значимы.


**Вывод:**

Гипотеза о том, что вероятность возврата кредита в срок зависит от семейного положения обоснована.

Клиенты, не оформившие своё семейное положения официальным браком имеют на 20-30 % большую долю просроченных кредитов, по сравнению с людьми, прошедшими официальную регистрацию брака (абсолютные вероятности просрочки для этих двух больших групп 9 % и 7 % соответственно).

Различия вероятностей просрочки внутри групп оформлявших и не оформлявших брак статистически НЕ значимы.

[Оглавление](#0-bullet)

<a class="ancor" id="3-bullet"></a>
#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Посмотрим распределение средней просрочки от категории дохода

In [32]:
debt_data_by_group("total_income_category")

Размеры групп клиентов с total_income_category = ['A' 'E'] слишком малы для анализа и не будут использоваться.


group_size group_size_percent  debt_fraction  \
total_income_category                                                 
B                            5014              23.51       0.070602   
C                           15921              74.64       0.084982   
D                             349               1.64       0.060172   

                      relative_debt_percent  
total_income_category                        
B                                    -16.92  
C                                      0.00  
D                                    -29.19

Размеры групп клиентов с total_income_category = ['A' 'E'] слишком малы для анализа.

Количество клиентов с доходами категории D также довольно низкое, оставим их для анализа, но будем иметь ввиду, что данные для этой категории не особо надёжны.

Средняя просрочка значительно различается для всех категории клиентов.

Наиболее надежными является клиенты с доходом категории D (доходы от 30 до 50 тысяч), на 17 % менее надёжными являются клиенты категории B (доходы от 200 тысяч до 1 миллиона) и самыми ненадёжными (на 29 % по сравнению с категорией D) являются клиенты категории C (доходы от 50 до 200 тысяч).

**Вывод:**

Гипотеза о том, что вероятность возврата кредита в срок зависит от дохода обоснована для клиентов с доходами от 30 тысяч до 1 миллиона.

Наиболее надежными является клиенты с доходами от 30 до 50 тысяч (средняя доля просрочки 6 %), менее надёжными являются клиенты с доходами от 200 тысяч до 1 миллиона (средняя доля просрочки 7 %) и самыми ненадёжными являются клиенты c доходами от 50 до 200 тысяч (средняя доля просрочки 8.5 %).

Однако, стоит отметить, что для клиентов с доходами от 30 до 50 тысяч, оказавшихся самыми надёжными, количество данных относительно невелико (349 записей), поэтому лучше проверить достоверность результатов на большем датасете.

Для клиентов в доходами менее 30 тысяч и более миллиона достоверных выводов сделать невозможно ввиду малочисленности таких клиентов (0.22 % от общего количества клиентов).

[Оглавление](#0-bullet)

<a class="ancor" id="4-bullet"></a>
#### 3.4 Как разные цели кредита влияют на его возврат в срок?

Посмотрим распределение средней просрочки от цели взятия кредита

In [33]:
debt_data_by_group("purpose_category")

group_size group_size_percent  debt_fraction  \
purpose_category                                                         
операции с автомобилем          4279              20.06       0.093480   
операции с недвижимостью       10751              50.40       0.072551   
получение образования           3988              18.70       0.092528   
проведение свадьбы              2313              10.84       0.079118   

                         relative_debt_percent  
purpose_category                                
операции с автомобилем                    0.00  
операции с недвижимостью                -22.39  
получение образования                    -1.02  
проведение свадьбы                      -15.36

Во всех категориях достаточное количество клиентов

Результаты разделились на 2 группы.

Клиенты, берущие кредит на операции с автомобилем и на получение образования оказались наиболее проблемными заёмщиками с более, чем 9 % средней просрочки. Клиенты, берущие кредит на проведение свадьбы и операции с недвижимостью имеют на 15 % и 22 % меньше просрочек.

При этом Статистические различия между клиентами, оформляющими кредит на операции с недвижимостью и проведение свадьбы не очевидны. Проверим статистическую значимость.

In [34]:
statistical_significance(
    data,
    "purpose_category",
    "debt",
    0.05,
    "chi_squared",
    ["проведение свадьбы", "операции с недвижимостью"],
)

Для клиентов с purpose_category = проведение свадьбы и операции с недвижимостью  p-значение = 0.2926.  Различия статистически НЕ значимы.


**Вывод:**

Гипотеза о том, что вероятность возврата кредита в срок зависит от цели выдачи кредита обоснована.

Клиенты, берущие кредит на операции с автомобилем и на получение образования оказались наиболее проблемными заёмщиками с более, чем 9 % средней просрочки. Клиенты, берущие кредит на проведение свадьбы и операции с недвижимостью имеют 15-22 % меньше просрочек (различия между "свадьбой" и "недвижимостью", а также между "автомобилем" и "образованием" статистически НЕ значимы).

[Оглавление](#0-bullet)

<a class="ancor" id="5-bullet"></a>
#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 

Пропуски в исходных данных встречаются только одновременно для колонок с общим доходом и количеством отработанных дней. Маловероятно, что такой характер пропусков мог возникнуть из-за случайного технического сбоя или человеческого фактора.

Вероятно, данные собирались из нескольких источников (отеделений банка?), и в одном из таких источников либо в анкете отсутствовали эти графы, либо произошла техническая ошибка при получении данных из этого источника.

Рекомендуется проверить источники данных и выяснить не принадлежат ли все пропуски к одному источнику данных.

[Оглавление](#0-bullet)

<a class="ancor" id="6-bullet"></a>
#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 

Медианные значения менее подвержены влиянию выбросов (экстремальных значений) и, в целом, лучше описывают усредненную ситуацию в исследуемом множестве.

Впрочем, медиана тоже не идеальна и бывают ситуации, когда и она может неадекватно описывать распределение, пример ниже.

In [35]:
ser = pd.Series([0] * 52 + [100] * 48)
ser.median()

np.float64(0.0)

В данном примере медиана имеет нулевое значение, что очень плохо описывает данный датасет.

Если есть возможность, то более надежно предварительно оценить набор численных данных методом descrive()

In [36]:
ser.describe()

count    100.000000
mean      48.000000
std       50.211673
min        0.000000
25%        0.000000
50%        0.000000
75%      100.000000
max      100.000000
dtype: float64

[Оглавление](#0-bullet)

<a class="ancor" id="7-bullet"></a>
### Шаг 4: общий вывод.

**Предобработка данных**

Предобработка данных выявила несколько проблем в данных:
* Аномальные значения в данных о рабочем стаже и количестве детей

     У некоторых клиентов в графе "количество детей" значатся значения -1 и 20. Количество таких записей незначительно (123), поэтому они не учитвылись при анализе.

    Данные по рабочему стажу содержат отрицательные и аномально большие (несколько сотен лет) значения. Отрицательные значения были заменены их абсолютным значением, но дальнейшая корректировка не проводилась, так как этот тип данных не использовался в дальнейшем анализе.
    

* Пропущенные значения в данных о рабочем стаже и общем доходе клиентов

    Пропущенные данные клиентов были заполнены медианными значениями для их типа занятости величинами.
    

* Дубликаты в данных - явные и неявные

    Неявные дубликаты содержатся в графе об образовании и связаны с написанием слов в различном регистре. Все слова в этой графе были приведены к строчному регистру.

    Данные содержат явные дубликаты (71 запись), которые были удалены.

Была проведена категоризация данных:
* Были выделены категории клиентов по уровню дохода:
  * 0–30000 — `'E'`;
  * 30001–50000 — `'D'`;
  * 50001–200000 — `'C'`;
  * 200001–1000000 — `'B'`;
  * 1000001 и выше — `'A'`.
  
* Были выделены категории целей получения кредита:
  * операции с автомобилем;
  * операции с недвижимостью;
  * проведение свадьбы;
  * получение образования.
  
**Исследование доли просроченных кредитов**

Была изучена зависимость доли просроченных кредитов от различных факторов:
* Количество детей
* Семейное положение
* Уровень дохода
* Цель получения кредита

Во всех случаях удалось выявить статистически значимые различия в количестве просроченных кредитов. При этом во всех случаях доля просроченных кредитов изменялась в диапазоне 6-9.5 %.

Для клиентов с детьми достаточное количество информации имеется только для клиентов с 2 и менее детьми. Для таких клиентов факт наличия хотя бы одного ребенка поднимает вероятность просрочки сразу на 20 %.

Люди, побывавшие в гражданском браке, на 20-30 % более надёжны людей, никогда не оформлявших своё семейное положение.

Влияние уровня дохода удалось достоверно установить только для клиентов с доходами категорий B и С (от 50 тысяч до 1 миллиона). Для этих категории клиенты с высоким доходом более надёжны.

Цель кредита на вероятность просрочки влияет следующим образом: Клиенты, получающие кредит на проведение свадьбы и операции с недвижимостью имеют на 15-22 % меньше просрочек, чем клиенты, имеющие своей целью покупку автомобиля или получение образования.

Таким образом, наиболее **надёжными** заёмщиками являются клиенты без детей, побывавшие в официальном браке, с общим доходом от 200 тысяч до 1 миллиона и получающие кредит на операции с недвижимостью или проведение свадьбы.

Наиболее **рискованными** заёмщиками являются клиенты с детьми, никогда не оформлявшие своё семейное положение, с доходом от 50 до 200 тысяч и получающие кредит на получение образования или покупку машины.


**Рекомендации**
* Предобработка данных выявила большое количество проблемных данных в графе общего стажа клиентов, поэтому рекомендуется проверить регламент работы с этим типом данных.
* Все пропуски в данных встречаются парами в графе о стаже и общем доходе. Вероятно, данные собираются из нескольких источников (отеделений банка?), и в одном из таких источников возникла проблема. Рекомендуется проверить источники данных и выяснить не принадлежат ли все пропуски к одному источнику данных.
* Наблюдается недостаточное количество данных для клиентов с 3 и более детьми, а так же для клиентов с доходом ниже 50 тысяч и более 1 миллиона. Рекомендуется собрать больше данных по этим категориям клиентов, чтобы можно было получить для них достоверные результаты.